### ISSUE TITLE
Status of mainline kernel support

### ISSUE THREAD

**Issue by sumpfralle** [gnubee-git/GnuBee_Docs#75]
I guess, it is of vital importance for the longevity of this project (and the hardware) to bring support the peripherals of GnuBee devices to the upstream kernel development tree. Otherwise we will end up with an outdated kernel (and thus more electronic landfill) somehwen ...

Since this is a not a short-term task, I propose to use this ticket for collecting the status of the ongoing mainlining efforts.

Could someone please start with summarizing the current state?
Thank you!

**Comment by neilbrown** [comment]
https://github.com/neilbrown/linux/commits/gnubee/v4.15  mostly works.  I just today noticed that only the first SATA controller works.  Might fix that tomorrow.
I've posted some patches for inclusion in drivers/staging: https://lkml.org/lkml/2018/3/14/1035  Will repost with some improvements tomorrow.

**Comment by sumpfralle** [comment]
Thank you for your quick response, for all the work you have put into mainline support and for the entertaining [lwn article](https://lwn.net/Articles/743609/) you wrote.
Your progress is way better, than I expected - this revives my positive emotions for the GnuBee platform. Thank you!

**Comment by Adirelle** [comment]
@neilbrown Have you hit [this bug](https://github.com/gnubee-git/GnuBee_Docs/issues/54) with the ethernet driver ?

Right now, I have a job that checks the network connectivity every minute and reboot the GB if it fails...

**Comment by alethiophile** [comment]
My experience is also that when it's booted into the Linux kernel, the network crash problem goes away. I see the problem when the board is in any of the three states uboot prompt/initramfs prompt/halted, but powered on.

**Comment by Adirelle** [comment]
Weird. It always happens on booted kernel (the librecmc-based one, provided by GB) and does not take the switch down. The GB still have ICMP (e.g. ping works) but the TCP and UDP become unusable. That how  the jobs detects a failure : if it can ping a fixed IP but cannot connect to its echo service, it causes the system to reboot.

**Comment by neilbrown** [comment]
Maybe there are two different bugs here - one that kills the switch and one that kills TCP but leaves ICMP working....

**Comment by xvybihal** [comment]
@Adirelle I also hit bug #54 with this mainline kernel. GnuBee PC1 is sitting here on my desk, working "fine" (quotes because of ping shown later). But when I come back some hours/day(s) later, I can not connect to it - the time when it becomes unavailable via network is random, as far as I can tell. I Have to poweroff and power on again, nothing else worked for me ((un)plug cable).

```
gnubee ~ # uname -a
Linux gnubee.jvi.cz 4.15.12+ #3 SMP Wed Apr 4 15:29:16 CEST 2018 mips GNU/Linux

gnubee ~ # cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 9 (stretch)"
NAME="Debian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

gnubee ~ # cat /etc/network/interfaces
auto lo
iface lo inet loopback

auto eth0
iface eth0 inet static
	address 172.16.202.254
	netmask 255.255.0.0
	gateway 172.16.whatever

```

Strange thing is, that when I run ping long enough, I see that every other while it takes gnubee quite long time to respond (100+ ms). I can **deffinitely** rule out faulty switch or cable.

https://gist.githubusercontent.com/xvybihal/b403c3c2677e423f3aea73feb8255a91/raw/d137ab85de894e2d5e3cb5bb9e2ffae16d023231/Gnubee%2520ping


Unfortunately I do not have UART cable to check what is going on, when GnuBee is not available via network.


This "thing" is nothing new, it was acting this way from the begining, with every kernel provided, and Deabian installed. Makes GnuBee pretty unusable for me, if I have to restart it several times a day.

**Comment by neilbrown** [comment]
Thanks for the detailed bug report.  I love details.
What is happening on the gnubee before it starts failing?
Is it just sitting there idle?  Is it initiating network traffic itself?  It is responding to requests?
Are there lots of requests, or occasional? or few?  What protocol or protocols?
I've started a job fetching a 10K file over http every 2 seconds.  I'll see if it is still working in the morning.

**Comment by Adirelle** [comment]
It does not seem to be linked to network activity. Stopping the transmission daemon to reduce the bandwidth and number of connections does not prevent it to happen (neither postponing it).  And I have downloaded an Ubuntu DVD torrent over a 1Gb link without any issue.

**Comment by neilbrown** [comment]
So what *is* your gnubee doing when the problem happens?  How do you first notice it?  How long after boot does it typically fail?  <1hr?  <12hrs?
If I can reproduce the problem it is very likely that I can fix it
If not ... there are some kernel messages in that other bug which might be enough of a hint.  Maybe I'll try copying a bigger file.

**Comment by xvybihal** [comment]
In my case, the GnuBee is doing nothing when it happens. What do I mean by nothing? The system installed is clean, no additional network daemons or services are installed. No network transfers are happening (except of time sync, and common stuff in the base system). I do not use it, it just sits here and doing nothing. The time when it happens is just random. Sometimes its hours, sometime day or two. 
Next time it happens, I will try to take out the SD card and copy some log here, thats probably the only thing I can do, without UART cable (which I am going to enquip myself with soon).

**Comment by xvybihal** [comment]
Looking at the log/messages, there might be something helpful. Uploading whole file - it looks pretty similar with what @Adirelle posted [messages.tar.gz](https://github.com/gnubee-git/GnuBee_Docs/files/1906029/messages.tar.gz)

**Comment by Adirelle** [comment]
> So what is your gnubee doing when the problem happens? How do you first notice it? How long after boot does it typically fail? <1hr? <12hrs?

Well, it seems totally random. The gnubee could be idle or I could be using it. Right now, it is running the following services : kernel nfs server, dropbear, dms (a upnp media server), ntpd, mysqld, transmission-daemon, ypbind, and there is a job running rdiff-backup once a day.

Sometimes it can be ok a whole week, sometimes only a few hours. I have the impression that the uptime rarely gets over 48 hours.

> If I can reproduce the problem it is very likely that I can fix it

I understand that -- I am a developer myself -- but unfortunately I could not find something to trigger the bug. I have only got the kernel error message.

I have a script that reboots it within one minute after it happens. It could dump the result of some commands in a file before rebooting, if you had some suggestions.

**Comment by neilbrown** [comment]
Thanks for posting the whole log file (did I say that I love details?)
The problem appears to be that interrupts get disabled on one of the CPUs - presumably CPU0 as that gets all device interrupts on my gnubee..  This is reported by RCU at timestamp 2179 which should be about 21 seconds after it happened.  Then at timestamp 6317 (about 1 hour later) the network watchdog complains.
With interrupts disabled on one CPU, the network card will become unreliable.  As polling is sometimes used you can still get some traffic through, but it won't be fast.  A 'ping' probably works because it sends a packet every send, and only wants one in reply.  If it checks for incoming packets whenever it transmits a packet (quite likely), it will appear to work normally.
The next question is: why do interrupts get disabled? Two things might be useful.
1/ @Adirelle if you could get your script to "echo t > /proc/sysrq-trigger" when the problem is detected, that might help.  It should write stack traces for all process to the kernel log.  Seeing those would be most helpful.
2/ @xvybihal If you could rebuild your kernel with LOCK_DEP enabled, that might produce more useful info (I hope).  You need to enable CONFIG_PROVE_LOCKING and CONFIG_DEBUG_LOCKDEP and you may as well add CONFIG_PROVE_RCU.  Then if it happens again, collect the logs the same way that you did before.
Thanks.

**Comment by neilbrown** [comment]
I have found a smoking gun.... I don't know if it is *the* smoking gun.
I went looking for places in the code which disable interrupts without clearly re-enabling them.  I didn't expect to find any - that would be too easy.  I found two!
One is almost certainly of no consequence - it always happens very early during initialization, so somethings must re-enable them.
The other is in the network code.  I don't understand the code enough to know when it happens, but it looks like it is in response to some event like the cable being pulled (though I tried that and it doesn't trigger anything).  The same bug is in the 4.4 kernel code
I've push out a fix to my gnubee/v4.15 kernel branch.  Please test in you can.

**Comment by neheb** [comment]
@neilbrown Have you tried patching the stock Mediatek MMC driver to add support for mt7621? I tried and failed. Something about missing pinctl. I think I needed to edit the dts file...

This was my basis: https://github.com/jonpry/openwrt_mt7688/commit/a85e6d99899f3dc1204cd5bfba944e17bfa6178f

**Comment by neilbrown** [comment]
> @neilbrown Have you tried patching the stock Mediatek MMC driver to add support for mt7621?

No I haven't.  As MMC is currently working, that driver is not a priority for me.  My current priorities are roughly:
- reboot - kernel currently leaves NOR flash in an inconsistent state
- SATA access seems really slow - 10MB/sec!
- network switch - enable VLAN support so different ports can be on different subnets
- 2nd network interface: the SOC has 2 interfaces to the switch, only one works at present
- crypto engine
- PCI driver has horrible hacks to select correct interrupt

So I won't be focusing on MMC for quite a while.  I'd be very happy for you or anyone else to dig into it and ask questions.  If you have specific focused questions, I'd be happy to share any expertise I might have.  I'd suggest opening a separate issue for each driver.

**Comment by Adirelle** [comment]
@neilbrown

Compiled your v4.15 and it boots (which is already something since I was not sure about what I was doing) but :
* the initramfs does not find my root partition, since it lies on a md RAID1 array.
* it messes with the attached switch, probably because its own switch is not configured.

Your kernel branch might or might not fix some bugs, but as long as it works, I prefer having a recent version quite easy to build over the ones from libreCMC/LEDE. I have tried

**Comment by Adirelle** [comment]
Got it working ; I was also missing some other modules. I will let you know if the network lockup happens again.

**Comment by neilbrown** [comment]
Another convert - hurray :-)
I've merged your patch - thanks.

**Comment by Adirelle** [comment]
Ok. So on the bad news:
* clock skew is back again (like https://github.com/gnubee-git/GnuBee_Docs/issues/49#issuecomment-347436829).
* performance seems lower than with the provided kernel, but I am not sure why.

I will try to compare kernel settings with the ones from LibreCMC.

**Comment by Adirelle** [comment]
@dgazineu Is there a way to flash the firmware from a running Linux ? (to shorten the whole "put your image on a USB stick, plug it on the GB, reboot, let uboot flash the firmware, remove the USB stick, reboot" cycle).

**Comment by neilbrown** [comment]
> clock skew is back again (like #49 (comment)).

You need the cpuclock clock-frequency in arch/mips/boot/dts/ralink/gbpc1.dts to match the value set by the u-boot that you have installed.  I have a u-boot that configures 900MHz so I set clock-frequency to 900000000.

> performance seems lower than with the provided kernel, but I am not sure why.

I just discovered that large kernel modules hurt performance.  I rebuilt with CONFIG_XFS=y (instead of =m, in O/.config) and filesystem throughput is a lot faster.  I've pushed an update for the defconfig file.

>  Is there a way to flash the firmware from a running Linux ?

Probably, using flash_erase and flash_cp from mtd-utils.  I haven't played much with them - be careful.
I set up my linux desktop as a tftp server and test kernels using tftpboot.  It is fairly painless.

**Comment by Adirelle** [comment]
> I have a u-boot that configures 900MHz so I set clock-frequency to 900000000.

Ah, I think mine is configured at 880Mhz. It seems there were some discrepancy between shipped u-boot and the kernel updates that were provided later.

> I set up my linux desktop as a tftp server and test kernels using tftpboot. It is fairly painless.

To be sure I understand you right : once the new image is available through TFTP, you use the provided u-boot menu to download it and flash it. Or do you run the kernel without flashing it (which would be ideal) ?

**Comment by neilbrown** [comment]
flash_erase works as expected.  flash_cp doesn't.  Something wrong with the spi driver...

**Comment by neilbrown** [comment]
I don't need to flash the kernel to test it.
The script I use to build the kernel finishes with

> cp O/arch/mips/boot/uImage.bin /srv/tftpboot/GB-PCx_uboot.bin
> echo 'tftpboot;bootm 80200000'

I reboot (or power-cycle) the gnubee and press '4' repeatedly during the early messages.  Once is probably enough, but more doesn't hurt.
When I get the prompt I cut/paste that last message printed by the script.
For this to work you at least need to set serverip in your u-boot environment
e.g.

> setenv serverip 192.168.1.4
> saveenv

You might also want to set, or at least check (printenv) bootfile, and ipaddr

**Comment by Adirelle** [comment]
> echo 'tftpboot;bootm 80200000'

Er... from where have you got the 80200000 ?

**Comment by neheb** [comment]
A small note. It seems the locking changes uncovered several other problems with the platform in general. One example: https://github.com/openwrt/openwrt/commit/173d93ea23dabf4f08228dc2aaf70a004cf4141c

There are several more with the wifi driver which are not applicable to the GnuBee.

Good to see this improved!

**Comment by neilbrown** [comment]
BTW, the tftpboot command doesn't need the magic number 80200000.
That is the default load address for tftpboot, but also the default address for bootm, so `tftpboot;bootm` is sufficient.

That openwrt patch is already in the upstream mtk eth driver, so we would have got it eventually.  It mostly just a fix for a false-positive warning though.
Certainly doesn't hurt to monitor what is happening in openwrt to see if anything might be useful - thanks.

**Comment by neheb** [comment]
Btw has anyone noticed data corruption on the drives after a days worth of uptime? I've noticed this on multiple MIPS devices since kernel 4.9 on OpenWrt but have avoided reporting to the kernel folks since it could very well be due to some OpenWrt patch.

Easiest way to detect is to use btrfs as it detects silent data corruption and floods dmesg.

**Comment by alethiophile** [comment]
I have seen significant data corruption on both hard drives and SD
cards. It's fairly annoying and prevents me from getting much done.

I attributed this originally to the age of the drives; I'm using
hardware that I had lying around. Haven't yet tested with known-good
stuff. If it is a problem with the Gnubee (software or hardware), it
definitely needs fixing.

On Wed, Apr 18, 2018 at 03:56:42PM +0000, Rosen Penev wrote:
> Btw has anyone noticed data corruption on the drives after a days worth of uptime? I've noticed this on multiple MIPS devices since kernel 4.9 on OpenWrt but have avoided reporting to the kernel folks since it could very well be due to some OpenWrt patch.
> 
> Easiest way to detect is to use btrfs as it detects silent data corruption and floods dmesg.
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/gnubee-git/GnuBee_Docs/issues/75#issuecomment-382437550

-- 
Tom Hunt

**Comment by Adirelle** [comment]
I have not noticed any corruption with 4.4 kernel. With 4.15, systemd-fsck complains about the ext4 FS in my mirrored logical volume (mirror LVM2) but not about the root on /dev/md1. When I run fsck on the former, it does not find nor fix any error. The disks were brand new.

**Comment by neilbrown** [comment]
I've run an fio test that explicitly looks for corruption and it ran for over 24 hours and reported nothing.  This was on a recent 4.15 kernel.  I might try running another test - or several at once.
If anyone does experience data corruption, please report exact kernel version, and what clock speed the CPU is configured for - over-clocking could cause problems.  Also report details of the corruption. e.g. output of fsck.
I think you can get clock speed with "dmesg | grep sys_freq".  I have 90000000.
I think I'll open a new issue to gather reports in.

**Comment by neheb** [comment]
This problem began since kernel 4.9. Maybe earlier but those kernels are EOL.

This does not just impact mt7621. OpenWrt's ar71xx platform is also affected. This indicates a MIPS specific problem. ARM and x86 are not affected.

This is a frustrating issue. At times, I have compiled firmware that seems to be working but the next one breaks.

This issue may be easier to trigger on low RAM devices. I'd have to test on my 32MB one.

**Comment by Adirelle** [comment]
@neilbrown  About the modules: I am wondering about what are the absolutely required drivers for the GnuBEE and which are optional in the default config for gb1. I would like to compile the former in the kernel, and compile the latter as modules or disable them.

**Comment by neheb** [comment]
An OpenWrt user discovered that network lockups may have something to do with flow control, so there's that.

Then again, we are talking about different drivers at this point...

**Comment by xvybihal** [comment]
> PS: on a side note, no network lockups (thus reboot) with the current kernel for a week. This is promising.

I have just up 4 days, 22 hours, 21 minutes, but I agree, by this time it usualy presented itself with no network connection, so fa so good, except onel ittle thing:
```
gnubee ~ # dmesg | grep 'mtk_soc_eth 1e100000.ethernet eth0:' | egrep "port 4 link down| port 4 link up" | wc -l
129
```
The link is going up and down from time to time, I am not sure if that's normal, but I would not say so.

**Comment by neilbrown** [comment]
> The link is going up and down from time to time, I am not sure if that's normal, but I would not say so.

Probably not normal.  The ethernet driver was in a fairly sad state when I got it - lots of work needed to make it work at all.  It is possible I broke something and certain that I didn't fix some things.  It's getting close to the top of my list of things to look at - after I get writing to flash working and streamline my process for building firmware images a bit.

**Comment by Adirelle** [comment]
> Is that helpful at all?

It is. Thanks.

> The link is going up and down from time to time, I am not sure if that's normal, but I would not say so.

The link seems pretty stable there. It only disconnected two times a week ago, and I might be for a real cause (like myself unplugging then plugging the cable, I cannot remember).

**Comment by neilbrown** [comment]
I've pushed out an update to gnubee/v4.15 which:
- fixes the data corruption bug (the network bug was already fixed)
- fixes writes to flash
- uses the configured MAC address for the network, instead of a random one (you need to rebuild the initramfs for this).
- updates to latest stable: 4.15.18

Now that writing to flash works, you can install new firmware with

    flashcp -v gnubee.bin /dev/mtd3

providing you have mtd-utils installed.

**Comment by Adirelle** [comment]
Built, tested, written to flash using flashcp, rebooted. Everything seems fine.

**Comment by Adonai** [comment]
Trying to build it on GB-1 itself: No luck, lzma says it can't allocate memory, 512 MiB seems not sufficient for the task

Trying to build it by cross-compiling: all good, extracted initramfs, compiled binutils, gcc-7.2.0, kernel, installed modules into initramfs, made uimage, flashed with USB stick - hmm, not booting.

Maybe it's my rootfs is to blame? It's the same one I used with 3.10 kernel.

**Comment by neilbrown** [comment]
> hmm, not booting.

What exactly does this mean?  Do you have a serial console and see no messages, or do you not have a serial console and see no flashing lights, or is it no serial console and you cannot ping/connect?
If the evidence of not booting is a lack of network connection, then it is probably that  network device names have changed.  In think with 3.10 you have "eth0.1" and "eth0.2" or something like that.  With mainline you just have "eth0".  If you can plug the uSD card into a different computer and edit /etc/network/interfaces and make that change, it might be worth another try.

**Comment by Adonai** [comment]
@neilbrown I don't yet have USB_2_UART to verify.
I guess you're right about network, will try to boot again today.

**Comment by Adonai** [comment]
@neilbrown you were right, it was network.
Works like a charm.

**Comment by neilbrown** [comment]
> I tried to run rsnapshot to create a recent backup of my desktop, but rsync ended up getting OOM killed,

Interesting... some sort of kernel memory leak I guess.
If you can trigger it again, it would be good if you could provide the contents of /proc/slabinfo when the machine is OOM.  I might not need that, but it will probably help.
Thanks.

**Comment by hvontres** [comment]
I tried to reproduce the issue tonight, but the job ran fine. I am guessing that the combination of network traffic/ disk activity might have caused the memory leak. Please let me know if there is anything else I can do the help diagnose this.
The disk in question is a raid10 array with an LVM volume on top of it.

**Comment by hvontres** [comment]
I tried to reproduce the issue tonight, but the job ran fine. I am guessing that the combination of network traffic/ disk activity might have caused the memory leak. Please let me know if there is anything else I can do the help diagnose this.
The disk in question is a raid10 array with an LVM volume on top of it.

**Comment by neheb** [comment]
Does valgrind work on mainline? In OpenWrt i'm getting:

==1804== Invalid read of size 4
==1804==    at 0x48C0ED4: free (in /usr/lib/valgrind/vgpreload_memcheck-mips32-linux.so)
==1804==    by 0x40865B4: ??? (in /lib/libc.so)
==1804==  Address 0x1e0d8 is not stack'd, malloc'd or (recently) free'd

On the topic of memory leaks, downloading a torrent with transmission ends up causing the kernel to kill it unless I add a swap device. It also seems that memory leaked by transmission doesn't seem to be freed as I still get OOM errors when logging on with ssh after disabling swap. Not sure if valid in mainline.

**Comment by neilbrown** [comment]
> Does valgrind work on mainline?
Nope.  I try "valgrind ls -l" and get a segmentation fault.
The same thing happens on my 4.4.131 kernel.  Maybe some needed CONFIG option is missing.

> On the topic of memory leaks,...
I'm not surprised that transmission gets killed - it probably uses too much memory.  Probably safest to leave swap enabled I guess...  You could look in /proc/meminfo and /proc/slabinfo to see if there is any sign of excess memory usage ... though they probably take a bit of expertise to understand them fully.

**Comment by neheb** [comment]
One thing I noticed is that top lists a single transmission-daemon instance whereas htop shows 4.

/proc does not show the extra three that get listed by htop. I wonder if limiting the CPUs to 1 will fix this...

**Comment by neilbrown** [comment]
htop is probably showing threads while top is showing processes.
If you have the pid as $PID, then look in
  /proc/$PID/task
There will be one directory for each thread.

NeilBrown

**Comment by xvybihal** [comment]
This seems to be transmission daemon problem:
https://github.com/transmission/transmission/issues/313

so 19. 5. 2018 v 9:05 odesílatel Rosen Penev <notifications@github.com>
napsal:

**Comment by neheb** [comment]
Yeah I looked at the code. It's threads. Seems this is a known issue...

**Comment by neheb** [comment]
I just did some testing in OpenWrt. Apparently in a default fashion, iperf results with the ethernet driver there give very inconsistent results. However, if I change the default smp_affinity from "f" to "8", there's a massive improvement in both consistency and performance. I have no idea if this can be done on a driver level.

**Comment by neilbrown** [comment]
Oooh, that's interesting. It's probable that something can be done in
the driver.  The first step would be understanding how the driver works
- I don't think that just hard-coding the affinity would be welcome.
So we are probably still a fair way from a proper solution, but this is
a valuable contribution towards it.

Thanks,
NeilBrown

**Comment by neheb** [comment]
I additionally did testing with nr-cpus=1 and nr-cpus=2. nr-cpus=1 gave me the most consistent results of ~540mbps +-10mbps. nr-cpus=2 gives me the same results. 

BUT, setting smp_affinity of the ethernet IRQ to either 2 or 4 (I forgot) such that /proc/interrupts show the IRQ being active only on CPU1 boost performance to around ~800mbps +- 10mbps.

I assume nr-cpus=2 limits mt7621 to the physical cores and keeps the SMT cores away.

Removing nr-cpus completely and setting smp_affinity to 8 gives me ~930mbps in the best case. I say best since that's around the limit of gigabit ethernet on iperf. I also say best as the variance comes back, but not as badly as before. ~770mbps IIRC was the worst case.

Without tweaking smp_affinity, iperf results go down to ~440mbps as as high as ~900mbps which is quite ridiculous.

**Comment by andreygursky** [comment]
@xvybihal:
> This seems to be transmission daemon problem:

For the issue with transmission I've found a workaround and described it here: https://github.com/transmission/transmission/issues/313#issuecomment-332543087.

This is not a memory leak.

**Comment by neilbrown** [comment]
Submitting patches is documented in the kernel tree, in Documentation/process/submitting-patches.rst.
It doesn't explicitly mention staging, but the same process applies.
Documentation/process/2.Process.rst has a bit about drivers/staging.
Prepare a patch and email to Greg Kroah-Hartman <gregkh@linuxfoundation.org>  and devel@driverdev.osuosl.org  and me (I'm listed in the TODO file).
get_maintainers.pl might give you a few too many addresses - definitely don't send to  John Crispin, he has said he isn't interested.  If you want you could just send a patch to me (neil@brown.name) and I'll tell you if you did anything seriously wrong, and exactly where it should be sent.

**Comment by xvybihal** [comment]
After some reboots lately, I noticed that my GB1 is getting different IPs from my DHCP instead of reserved lease. After little investigation I noticed that at every reboot the GB1 has new MAC. Is that by design? Why?

From kernel log:
`mtk_soc_eth 1e100000.ethernet: generated random MAC address d2:44:2f:d4:75:c1`

**Comment by neilbrown** [comment]
> After little investigation I noticed that at every reboot the GB1 has new MAC. Is that by design? Why?

The gnubee network hardware doesn't have a MAC address burnt into to, so one has to be loaded by software.
The libreCMC kernel (and probably others) have a hack in the kernel to read a mac address out of the flash memory and use that.  The devicetree file contains:

> &ethernet {
        mtd-mac-address = <&factory 0xe000>;
};

which says "read 6 bytes from offset 0xE000 in the "factory" partition of the flash, and use that for the mac address.
The upstream kernels don't have this.  However the initramfs that I provided with the kernels that I provide contains:

> addr=\`dd 2> /dev/null if=/dev/mtd2 bs=1 skip=57344 count=6 | od -t x1 -A none\`
> ip link set eth0 address \`echo $addr | tr ' ' :\`

in the init script.  This reads the same 6 bytes out of the same place in flash, and uses them to set the mac address for eth0.
So: what kernel are you running?  And what do you get if you run

> dd 2> /dev/null if=/dev/mtd2 bs=1 skip=57344 count=6 | od -t x1 -A none

??

**Comment by xvybihal** [comment]
> However the initramfs that I provided with the kernels that I provide contains:

oh I see, I made the initramfs long time ago, certainly before your commit c493c10, thats why I do not have this in place

> So: what kernel are you running? And what do you get if you run
I have your 4.15 kernel, and when I run that, I certainly do get what i need for MAC:
```
gnubee ~ # dd 2> /dev/null if=/dev/mtd2 bs=1 skip=57344 count=6 | od -t x1 -A none
 78 a3 51 ab b7 1c
```
I think I will just hardcore it into initramfs for my purposes. Thanks for additional info, appreciate it.

**Comment by vgiralt** [comment]
Hi! I was hit by the "frozen switch" bug that brought my whole house network :-( while copying the MMC partition contents to an SSD using tar after a failed dd (due to overseen space limitation), this led me into trying @neilbrown 's gnubee-pc1-900-4.4.144.bin, but this leaves me with no connection (I've edited the MMC based Debian configuration interfaces to replace eth0.1 with eth0) and the PC1 has a solid lower green led and upper led off (it flashed shortly once while booting). I do not have a USB-2-UART cable. I do have a USB-2-TTL from another project but do not know where to connect it :-(
Any suggestions? Thanks!

**Comment by vgiralt** [comment]
BTW, if anyone could give me a clue of the .bin file _innards_ I might be able to have a look inside and, maybe, I could debug further without disturbing here :-) I used to compile kernels back in the '90s, but now I _limit_ myself to managing big infrastructures with SANs, blades and VMs ;-)

**Comment by neilbrown** [comment]
For the usb-2-ttl, there are three pins on the underside near the reset-button and 3.5mm photo jack.  From front to back these are black, green, white - though there should be 'rx' and 'tx' labeling on the silk screen.

If you want to build your own firmware, I recommend https://github.com/neilbrown/gnubee-tools

**Comment by vgiralt** [comment]
Thanks @neilbrown for such a fast response! Actually I'm more interested in helping debug than setting up a cross compiling environment. The main reason for acquiring the PC1 is reducing the energy footprint of my "home datacenter" :-), I'm a bit short on time lately :-) but I prefer fully opensource things and the GnuBee seemed worth supporting. That's enough "noise".
Again, thank you. I had spotted the Tx RX pins you mention, but was not sure the would work. I'll try to attach the USB-2-TTL dongle I have and try to see what's really happening.
BTW, I had already perused your gnubee-tools project, but I somehow failed to really read teh boot process. Now, I've seen that your image _should_ find the MMC filesystem and boot into Debian.

**Comment by vgiralt** [comment]
On booting the MMC: It seems to boot (configures the network and does pings from a if-up-post.d script) but that seems to be all, no ssh access, no open ports ...
I resorted to installing Debian on a disk partition and all goes well, except it will not boot from a LV, as there seems to be no support for it in your (@neilbrown) init script :-(
Regarding LVM, I cannot use it once booted with your (@neilbrown) gnubee-pc1-900-4.4.144.bin, I get this error:
`` Incompatible libdevmapper 1.02.137 (2016-11-30) and kernel driver (unknown version)``
although I get a list of available volumes, I cannot activate the group, neither use the volumes.
I was ready to compile a firmware, but, the I got doubts about which modules I should activate for LVM to work properly.

**Comment by vgiralt** [comment]
Any information on network devices naming and configuration?
Booting with @neilbrown 4.4.144 mainline kernel into a Debian system, journalctl --system shows two active ports:
`Sep 05 18:07:50 gnubee-pc1.gnubee kernel: mtk_soc_eth 1e100000.ethernet eth0: port 4 link up`
`Sep 05 18:08:22 gnubee-pc1.gnubee kernel: mtk_soc_eth 1e100000.ethernet eth0: port 0 link up`
But there is only *one* MAC address,while librecmc and lede show *two*.
I've tried to configure IP addressing on eth0 to no avail, instead, I can config eth0.1 and add working IPv4 addresses on eth0.2, eth0.3 or eth0.4. IPv6 only works on eth0.1, but it makes sense given the single MAC address.
The working port is the 100 MB/s one, not the 1GB/s one.

**Comment by smurfix** [comment]
Meh. Any progress / anything I could do to help?

The 4.4 kernel that comes with the system starts getting too old to run Debian Testing these days.

**Comment by neilbrown** [comment]
I've been really busy for the last few months and had no time for the gnubee.  Things are settling down at last and I hope to start getting involved again.  My main focus will be getting the mainline kernel to work.
Just at the moment it doesn't even build, and when I fix those issues, various bits don't work.  One useful thing that other people could do would be to monitor patches going into the 'staging' tree and make sure they don't break things.  The people who are working to improve the code often aren't in a position to test it.
You could subscribe to  driverdev-devel@linuxdriverproject.org to see new patches, and test them even before they land in staging.  When a patch does cause a breakage, it can help a lot to report that with any error messages.  Even if they don't mean much to you, they can help someone else to figure out what is happening.

**Comment by smurfix** [comment]
@neheb Link?
@neilbrown Thanks. Will do.

**Comment by neheb** [comment]
@smurfix https://github.com/openwrt/openwrt/pull/1551

**Comment by neilbrown** [comment]
Just FYI, I have a near-mainline kernel working o gnubee again.
https://github.com/neilbrown/linux/tree/gnubee/v4.20

**Comment by neilbrown** [comment]
@neheb  I've added that fix from  openwrt/openwrt#1551 to my neilbrown/linux/gnubee/v4.20 - it appears to work, and means that we don't need to know the clock-speed at build time.  Excellent.

**Comment by neheb** [comment]
Fun little project: https://github.com/vschagen/mtk-eip93

From initial testing on the PC1, results are mixed. OpenSSL speed results:

Before:
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
aes-128-cbc       2125.04k     8559.05k    43661.36k   170023.38k  1672396.80k  1725235.20k

After:
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
aes-128-cbc       1931.52k    12866.13k    39646.61k   160035.84k  1898741.76k 10901913.60k

Driver seems stalled and not under active development. I also experience CPU stalls with it. For example, cryptosetup benchmark causes a CPU stall when it gets to AES128.

cryptsetup benchmark -c aes gives

Before: aes-cbc        256b         9.1 MiB/s         9.2 MiB/s
After: aes-cbc        256b        33.1 MiB/s        33.2 MiB/s

cryptsetup benchmark -c aes --key-size 128 causes a CPU stall.

The driver could probably use more optimization.

**Comment by neilbrown** [comment]
> Fun little project: https://github.com/vschagen/mtk-eip93
It would be great to get this into drivers/staging - more eyes on the code, more awareness, more testing (maybe).

**Comment by Adirelle** [comment]
I guess that without optimization, using hardware is still way faster than software.

**Comment by Adirelle** [comment]
@neilbrown: got this when trying to compile v4.20. Have I overlooked something ?

```
  GEN     Makefile
scripts/kconfig/conf  --syncconfig Kconfig
  GEN     Makefile
  Using .. as source for kernel
  CALL    ../scripts/checksyscalls.sh
  CHK     include/generated/compile.h
  CC      arch/mips/ralink/mt7621.o
  GEN     usr/initramfs_data.cpio.lzma
  CC      arch/mips/kernel/cmpxchg.o
  CC      arch/mips/mm/cache.o
  CC      arch/mips/kernel/cpu-probe.o
../arch/mips/ralink/mt7621.c:15:10: fatal error: dt-bindings/clock/mt7621-clk.h: No such file or directory
 #include <dt-bindings/clock/mt7621-clk.h>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[3]: *** [../scripts/Makefile.build:291: arch/mips/ralink/mt7621.o] Error 1
```

**Comment by neilbrown** [comment]
sorry, I forgot to "git add" a file.
I've fixed up the relevant patch and re-pushed.
As I rebased, you need to reset to an earlier version.
e.g.
  git reset --hard HEAD~10
  git pull  https://github.com/neilbrown/linux.git gnubee/v4.20

NeilBrown

On Mon, Dec 24 2018, Adirelle wrote:

> @neilbrown: got this when trying to compile v4.20. Have I overlooked something ?
>
...
> ../arch/mips/ralink/mt7621.c:15:10: fatal error: dt-bindings/clock/mt7621-clk.h: No such file or directory
>  #include <dt-bindings/clock/mt7621-clk.h>
>           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> compilation terminated.
> make[3]: *** [../scripts/Makefile.build:291: arch/mips/ralink/mt7621.o] Error 1

**Comment by Adirelle** [comment]
Two more: 

```
  GEN     Makefile
scripts/kconfig/conf  --syncconfig Kconfig
  GEN     Makefile
  Using .. as source for kernel
  CALL    ../scripts/checksyscalls.sh
  CHK     include/generated/compile.h
  GEN     usr/initramfs_data.cpio.lzma
  HOSTCC  lib/gen_crc64table
../lib/gen_crc64table.c:19:24: fatal error: linux/swab.h: No such file or directory
 #include <linux/swab.h>
                        ^
compilation terminated.
make[3]: *** [scripts/Makefile.host:90: lib/gen_crc64table] Error 1
```
and:

```
  AS      usr/initramfs_data.o
  AR      usr/built-in.a
  CC      drivers/watchdog/mt7621_wdt.o
../drivers/watchdog/mt7621_wdt.c:174:34: error: array type has incomplete element type ‘struct of_device_id’
 static const struct of_device_id mt7621_wdt_match[] = {
                                  ^~~~~~~~~~~~~~~~
../drivers/watchdog/mt7621_wdt.c:175:4: error: field name not in record or union initializer
  { .compatible = "mediatek,mt7621-wdt" },
    ^
../drivers/watchdog/mt7621_wdt.c:175:4: note: (near initialization for ‘mt7621_wdt_match’)
../drivers/watchdog/mt7621_wdt.c:174:34: warning: ‘mt7621_wdt_match’ defined but not used [-Wunused-variable]
 static const struct of_device_id mt7621_wdt_match[] = {
                                  ^~~~~~~~~~~~~~~~
make[4]: *** [../scripts/Makefile.build:292: drivers/watchdog/mt7621_wdt.o] Error 1
```

I'd report them to your [linux fork](https://github.com/neilbrown/linux) but you disabled the issues.

**Comment by xvybihal** [comment]
Very similar build error here:

```
  CC      drivers/watchdog/mt7621_wdt.o
../drivers/watchdog/rt2880_wdt.c:188:34: error: array type has incomplete element type
 static const struct of_device_id rt288x_wdt_match[] = {
                                  ^
../drivers/watchdog/rt2880_wdt.c:189:2: error: field name not in record or union initializer
  { .compatible = "ralink,rt2880-wdt" },
  ^
../drivers/watchdog/rt2880_wdt.c:189:2: error: (near initialization for 'rt288x_wdt_match')
../drivers/watchdog/rt2880_wdt.c:188:34: warning: 'rt288x_wdt_match' defined but not used [-Wunused-variable]
 static const struct of_device_id rt288x_wdt_match[] = {
                                  ^
make[3]: *** [../scripts/Makefile.build:292: drivers/watchdog/rt2880_wdt.o] Error 1
make[3]: *** Waiting for unfinished jobs....
  CC      drivers/usb/phy/of.o
  CC      lib/xarray.o
../drivers/watchdog/mt7621_wdt.c:174:34: error: array type has incomplete element type
 static const struct of_device_id mt7621_wdt_match[] = {
                                  ^
../drivers/watchdog/mt7621_wdt.c:175:2: error: field name not in record or union initializer
  { .compatible = "mediatek,mt7621-wdt" },
  ^
../drivers/watchdog/mt7621_wdt.c:175:2: error: (near initialization for 'mt7621_wdt_match')
../drivers/watchdog/mt7621_wdt.c:174:34: warning: 'mt7621_wdt_match' defined but not used [-Wunused-variable]
 static const struct of_device_id mt7621_wdt_match[] = {
                                  ^
make[3]: *** [../scripts/Makefile.build:292: drivers/watchdog/mt7621_wdt.o] Error 1
make[2]: *** [../scripts/Makefile.build:516: drivers/watchdog] Error 2
make[2]: *** Waiting for unfinished jobs....
```

**Comment by Adirelle** [comment]
> `../lib/gen_crc64table.c:19:24: fatal error: linux/swab.h: No such file or directory`

I have fixed this error by installing linux headers on the host. I find this a bit weird: the headers from current source should be used instead, isn't it ?

**Comment by smurfix** [comment]
Umm … no. `gen_crc64table` is run on the build host.

**Comment by Adirelle** [comment]
> Umm … no. `gen_crc64table` is run on the build host.

Yes, I know. This is why I am perplexed: if I did not install the linux-headers package on the build host (which would not be possible on a non-linux distro, btw), I would not be able to cross-compile the kernel from its sources.

**Comment by Adirelle** [comment]
> ```
> ../drivers/watchdog/rt2880_wdt.c:188:34: error: array type has incomplete element type

I disabled the rt2880 watchdog, considering it would probably not be in the rt7621 but I am not sure of that.

**Comment by Adirelle** [comment]
Maybe we should ping @neilbrown . ^^

**Comment by neilbrown** [comment]
The compile error in rt2880_wdt.c and mt7621_wdt.c is caused by 
Commit ac3167257b9f ("headers: separate linux/mod_devicetable.h from linux/platform_device.h")

We need to add 
#include <linux/mod_devicetable.h>
to both those files.  I'll submit a patch.
I suspect that mt7621_wdt.c is the driver to use if you want a watchdog, though I haven't tested it.
There might be problems using the watchdog together with the flash-memory as the flash controller needs to be reset for a reboot, and the hardware watchdog doesn't force a reset.  Something can probably be done as long as the flash controller isn't being used when the watchdog fires.

**Comment by neilbrown** [comment]
The problem with compiling gen_crc64table was caused by
Commit feba04fd2cf8 ("lib: add crc64 calculation routines")
and can be fixed by simply removing the 
#include <linux/swab.h>
from the source file.  I've posted a patch.

I've pushed out a new gnubee/v4.20 which contains these patches, and also updates to v4.20-final.

**Comment by xvybihal** [comment]
@neilbrown I successfully build your gnubee/v4.20 for GB1 and it booted, but without any network interfaces. Is that known/expected? When I build 4.15 with same defconfig, it works as expected, network including.
Anybody else tried it?

**Comment by neilbrown** [comment]
> @neilbrown I successfully build your gnubee/v4.20 for GB1 and it booted, but without any network interfaces. 

That is not expected.  Can you confirm that you have
CONFIG_NET_VENDOR_MEDIATEK_STAGING=y
CONFIG_NET_MEDIATEK_MT7621=y
CONFIG_NET_MEDIATEK_SOC_STAGING=y

in your .config ?
Does
  dmesg | grep eth

produce anything?

**Comment by xvybihal** [comment]
> That is not expected. Can you confirm that you have
> CONFIG_NET_VENDOR_MEDIATEK_STAGING=y
> CONFIG_NET_MEDIATEK_MT7621=y
> CONFIG_NET_MEDIATEK_SOC_STAGING=y

I only had CONFIG_NET_MEDIATEK_MT7621=y

> Does
> dmesg | grep eth
> 
> produce anything?
I can not check, because I flashed back the 4.15.18 which worked. I will try to build new image today with the suggested .config values.

Btw, in your gnubee/v4.20 branch, I can not see any gnubee defconfig, so I used the one from 4.15 and modified some stuff related to filesystems, fuse, etc. Maybe you forget to commit it, or we should use mt7621_defconfig?
Thanks

**Comment by neilbrown** [comment]
> Btw, in your gnubee/v4.20 branch, I can not see any gnubee defconfig, 

I've just pushed out an update which contains gbpc1_defconfig which is just the config I use.  Of course, one danger of using that is you only test the things I test, so you might not find bugs that I don't find...  
Have fun.

**Comment by Adirelle** [comment]
Speaking of that, I just compiled and booted the new version, with my own configuration file (see [there](https://github.com/Adirelle/gnubee-kernel/blob/master/gnubee1_defconfig)). It seems to boot fine, with the network, but it fails to activate a raid logical volume. I have checked that the device mapper is properly configured, after https://github.com/Adirelle/gnubee-kernel/blob/master/gnubee1_defconfig#L1407.

Here are the relevant error messages:
```
[   32.512176] device-mapper: table: 252:5: raid: Component device(s) too small
[   32.526398] device-mapper: ioctl: error adding target to table
[   33.176958] device-mapper: table: 252:5: raid: Component device(s) too small
[   33.191491] device-mapper: ioctl: error adding target to table
```

252:5 is the device number of the actual LV on v4.15. 

Here is the output of `lvs -a -o path,name,kernel_major,kernel_minor,devices,attr,size,sync_percent` with the v4.15 kernel:
```
  Path             LV              KMaj KMin Devices                           Attr       LSize   Cpy%Sync
  /dev/vgdata/data data             252    5 data_rimage_0(0),data_rimage_1(0) rwi-aor--- 925,00g 100,00
                   [data_rimage_0]  252    2 /dev/sda2(129)                    iwi-aor--- 925,00g
                   [data_rimage_1]  252    4 /dev/sdb2(1)                      iwi-aor--- 925,00g
                   [data_rmeta_0]   252    1 /dev/sda2(128)                    ewi-aor---   4,00m
                   [data_rmeta_1]   252    3 /dev/sdb2(0)                      ewi-aor---   4,00m
  /dev/vgdata/swap swap             252    0 /dev/sda2(0)                      -wi-ao---- 512,00m
```
Ouput of the same command with v4.20:
```
Path             LV              KMaj KMin Devices                           Attr       LSize   Cpy%Sync
  /dev/vgdata/data data              -1   -1 data_rimage_0(0),data_rimage_1(0) rwi---r--- 925,00g
                   [data_rimage_0]  252    2 /dev/sda2(129)                    Iwi-a-r-r- 925,00g
                   [data_rimage_1]  252    4 /dev/sdb2(1)                      Iwi-a-r-r- 925,00g
                   [data_rmeta_0]   252    1 /dev/sda2(128)                    ewi-a-r-r-   4,00m
                   [data_rmeta_1]   252    3 /dev/sdb2(0)                      ewi-a-r-r-   4,00m
  /dev/vgdata/swap swap             252    0 /dev/sda2(0)                      -wi-ao---- 512,00m
```

I am wondering what changed about DM raid that could cause this error. Glaring at `git log gnubee/v4.15..gnubee/v4.20` have not resulted in the expected enlightenment.

**Comment by neilbrown** [comment]
> [   32.512176] device-mapper: table: 252:5: raid: Component device(s) too small

That error message - and the check which leads to it - doesn't exist in 4.15.  It was added in 4.16 by commit 188a212df1f3a2d7ea9bb0fc0ab4173042c23470

Can you report the output of "dmsetup table" and "dmsetup info" on 4.15 when the array is running properly?

**Comment by Adirelle** [comment]
There you go: https://gist.github.com/Adirelle/667fba3b821fcceaefacf2ed595da4ad

**Comment by neilbrown** [comment]
Thanks.  That allows me to rule out some possible causes. I think it is a code bug.
Could you please edit include/linux/device-mapper.h in the 4.15 code and change

> static inline sector_t to_sector(unsigned long n)

to

> static inline sector_t to_sector(unsigned long long n)

i.e. change "long" to "long long".  Then rebuild and see if it works.
The size of your component device is 993211187200 bytes, which when stored in an "unsigned long" on a 32-bit arch is truncated to 1073741824.  to_sector() converts this to 2097152 sectors, which is much smaller than the array requirement of 1939865600 sectors - hence the error.
When you confirm that this fixes the problem, I'll post a patch.  If you would like to be acknowledged with a "Reported-and-tested-by:" tag, let me know and give me an email address.

**Comment by Adirelle** [comment]
> in the 4.15  code

1/ Should it not be in the 4.20 code, since it is not even checked in the 4.15 ?
2/ Does it only affect a check or should I expect some data corruption or other side effects with 4.15 ?

**Comment by Adirelle** [comment]
It fixes the error on v4.20. I made read-only fsck of the underlying filesystem and it seems ok. Should I test with v4.15 too ?

BTW, is there a way to build the initramfs and u-boot image in the path indicated by the "O" variable instead of the source tree ?

**Comment by neilbrown** [comment]
1/ Yes, 4.20.  Sorry.
2/ No data corruption.  You might get an error if you try to reshape an array or add a journal.

Thanks for testing and reporting - I'll send the patch upstream.
I doubt you would notice any difference if you made the same change in 4.15.

If you set GNUBEE_INITRAMFS_TREE in gnubee-tools/config to some other directory, the initramfs should be built there.  The u-boot image is always built in the O directory ($GNUBEE_KERNEL_OBJECTS) and then copied to $GNUBEE_BUILD_DIR.
Everything *should* be configurable in the 'config' file.

**Comment by Adirelle** [comment]
Thank you.

Here are more error messages (which does not seem to prevent anything from working, though):

```
[   11.599861] mt7621_gpio 1e000600.gpio: registering 32 gpios
[   11.615781] gpio gpiochip1: (1e000600.gpio-bank1): detected irqchip that is shared with multiple gpiochips: please fix the driver.
[   11.639095] mt7621_gpio 1e000600.gpio: registering 32 gpios
[   11.652240] gpio gpiochip2: (1e000600.gpio-bank2): detected irqchip that is shared with multiple gpiochips: please fix the driver.
[   11.675488] mt7621_gpio 1e000600.gpio: registering 32 gpios
```

```
[   11.775701] cacheinfo: Failed to find cpu0 device node
[   11.786392] cacheinfo: Unable to detect cache hierarchy for CPU 0
```

```
[   12.324852] ------------[ cut here ]------------
[   12.334100] WARNING: CPU: 2 PID: 1 at /home/user/gnubee/linux/drivers/mtd/spi-nor/spi-nor.c:3659 spi_nor_init+0x134/0x1d8
[   12.369010] enabling reset hack; may not recover from unexpected reboots
[   12.382345] Modules linked in:
[   12.388419] CPU: 2 PID: 1 Comm: swapper/0 Not tainted 4.20.0+ #11
[   12.400529] Stack : 00000123 00000000 00000000 80068b8c 00000001 811e3814 80640000 0000000b
[   12.417147]         807054b8 9bc81944 00000000 811e0000 80720000 00000001 9bc818d8 53260ce1
[   12.433766]         00000000 00000000 81220000 00000007 00000000 6465746e 00000123 00000000
[   12.450382]         00000000 00000122 81210000 79616d20 80720000 00000000 80740000 806db044
[   12.450402]         00000009 00000e4b 9bc81b88 00000000 00000000
[   12.450436]  802fca50 00000008 811e0008
[   12.450442]         ...
[   12.450448] Call Trace:
[   12.450472] [<8000c244>] show_stack+0x8c/0x130
[   12.450500] [<805c63a4>] dump_stack+0x94/0xd0
[   12.450514] [<80027d00>] __warn+0x10c/0x114
[   12.450524] [<80027d48>] warn_slowpath_fmt+0x40/0x64
[   12.450535] [<80375bf8>] spi_nor_init+0x134/0x1d8
[   12.450546] [<80378b88>] spi_nor_scan+0x8f8/0xa60
[   12.450562] [<80366d60>] m25p_probe+0x178/0x218
[   12.450573] [<8030b768>] really_probe+0x2cc/0x430
[   12.450595] [<803094a4>] bus_for_each_drv+0xac/0xcc
[   12.450604] [<8030b9ec>] __device_attach+0xbc/0x130
[   12.450616] [<8030a214>] bus_probe_device+0x3c/0xb0
[   12.450626] [<80307ef0>] device_add+0x494/0x5b0
[   12.450648] [<80390f08>] spi_add_device+0x148/0x1b0
[   12.450659] [<803919e8>] spi_register_controller+0x7a4/0x940
[   12.450679] [<8030d708>] platform_drv_probe+0x40/0x7c
[   12.450688] [<8030b768>] really_probe+0x2cc/0x430
[   12.450697] [<8030be98>] __driver_attach+0xb4/0x138
[   12.450707] [<803093a0>] bus_for_each_dev+0x6c/0xb0
[   12.450719] [<8030a5d8>] bus_add_driver+0x204/0x24c
[   12.450728] [<8030c7d8>] driver_register+0xd0/0x118
[   12.450738] [<80001638>] do_one_initcall+0x84/0x19c
[   12.450758] [<80753f2c>] kernel_init_freeable+0x248/0x250
[   12.450775] [<805e1fdc>] kernel_init+0x14/0x110
[   12.450784] [<80006838>] ret_from_kernel_thread+0x14/0x1c
[   12.450832] ---[ end trace b800848cea8dadd4 ]---
```

By the way, it seems the SATA drive configured for UDMA/133 despite a link at 6.0 Gpbs:

```
[   13.045238] ata2: SATA link up 6.0 Gbps (SStatus 133 SControl 300)
[   13.141553] ata2.00: ATA-10: ST1000LX015-1U7172, SDM1, max UDMA/133
[   13.154447] ata2.00: 1953525168 sectors, multi 16: LBA48 NCQ (depth 32), AA
[   13.262441] ata2.00: configured for UDMA/133
```

For reference, I used this [config file](https://github.com/Adirelle/gnubee-kernel/blob/99f1d00555224f76a95d099d9e21e1f3c779083c/gnubee1_defconfig). I have tried to disable stuff my GB1 do not need, to include always-used drivers and to enable as modules what I may need (like USB mass storage). I hope I have not disabled essential things.

**Comment by neilbrown** [comment]
The warning in spi-nor is annoying but harmless.
Some background can be read here: https://patchwork.ozlabs.org/patch/950299/
The hardware is not "broken" - but I didn't think it was worth fighting too hard.

I don't know much about ATA speeds so I cannot comment on that issue.   I doubt it is related to your config-file choices.

**Comment by Adirelle** [comment]
What about `detected irqchip that is shared with multiple gpiochips: please fix the driver.` ? Not that I have some use for the GPIO but could it cause a bug ? IIRC, there is only one GPIO available for user on the GB.

> I don't know much about ATA speeds so I cannot comment on that issue.

Well, no HD nor SDD reaches 6 Gb/s (aka ATA-600, 600 to be compared to the 133 of UDMA-133) but I would like to be sure the disk rate was not limited by the link. However, it seems the ST1000LX015-1U7172 hardly reaches 93 MB/s so that should not be an issue.

> I doubt it is related to your config-file choices.

Ok. While reviewing the drivers to enable, I was wondering about the hardware that is found or not on the GB-PC1. E.g., what is SPI used for and does the GB-PC1 use it ? Is it actually needed ? Same questions for I2C, ...

@neheb 
> Fun little project: https://github.com/vschagen/mtk-eip93

I am not familiar with hardware crypto engines. Could userland software (e.g. openss[lh] and the like) use the hardware crypto ?

**Comment by neilbrown** [comment]
SPI is used to access the Flash storage.
GPIO is used to drive the LEDs and to sense the push-button.
I don't think I2C is used.
Thanks for mentioning the irqchip thing - I'll look into it.
I believe most crypto libraries will use kernel-supported hardware when available, I'm fairly sure that includes openssl and I suspect that includes openssh.

**Comment by neheb** [comment]
The hardware only supports AES in CBC mode. SSH does not do CBC. dm-crypt defaults to XTS but can be configured to use CBC.

**Comment by Adirelle** [comment]
It seems the network bug (ping is working but not TCP nor UDP) is back. I have not identified what causes it and I have no suspicious message in dmesg or logs. It happens with both the 4.15.18 & 4.20 kernels. 
Trying to ifdown+ifup eth0 does not work but a reboot does.

**Comment by cmm** [comment]
@Adirelle do you use NFS (or Samba with sendfile() on)?

**Comment by Adirelle** [comment]
@cmm the kernel NFS server, yes. I do not know if sendfile is enabled. I removed samba a few week ago.

**Comment by cmm** [comment]
@Adirelle I started getting the network bug a _lot_ once my Gnubee started serving mostly 1080p media (instead of mostly 720p, where I was having those lockups maybe once a month); the problem disappeared completely once I stopped using NFS and moved to Samba-sans-sendfile().

it is my not-entirely-informed guess that the VFS/network interplay in the kernel is screwy, probably due to how the switch code does locking.  in fact, any in-kernel code that serves data streams through the switch is probably dangerous (I don't know if there is any apart from sendfile() & NSF, though).  the _actual_ problem here is that most SoC vendors just don't test such configurations -- most small NAS systems on the market don't do NFS at all, and the SoC in Gnubee is made for routers...

**Comment by Adirelle** [comment]
@neilbrown the initramfs init script made me a joke these days: as I was recompiling for the same kernel version, it refused to mount /lib/modules from the initramfs and used the /lib/modules instead, which was an issue since I changed the module settings. Is there something in the modules folder that could be used to known if the kernel builds are different ?

**Comment by neilbrown** [comment]
>  Is there something in the modules folder that could be used to known if the kernel builds are different ?

No.  For my gnubee-tools package, I create a 'stamp' file with the date when the modules were copied in, and compare that.
You could set CONFIG_LOCALVERSION_AUTO=y.  This adds part of the git hash of the top commit to the version so there is no risk of using old libraries with a newer kernel.
I've updated the gnubee1_defconfig in v4.15 to have this changed.

**Comment by neilbrown** [comment]
The mediatek network driver in mainline now supports the MT7621, so we don't need the drivers/staging driver.
It works with DSA support for the integrated switch.  I now have it working on both network interfaces on the PC1 and all three on the PC2.
So I'm now using mainline (5.1-rc2) on my main gnubee and will be building firmware with mainline kernels from time to time.  See the announcement in the google group.
I now consider the gnubee to be "fully supported" in mainline.  Though there is still a bit of work to do, it is mostly cleaning up the code and getting it moved out of drivers/staging (spi is almost there, and mmc probably isn't far away).

**Comment by neheb** [comment]
The MMC driver situation is the same as the Ethernet. The mainline mtd-sd can work with it with some modifications.

**Comment by smurfix** [comment]
Nice. Thanks for the work. Does mainline have a usable kconfig file? if not (or not yet …), where can I find one?

**Comment by vgiralt** [comment]
@neilbrown 
> I now consider the gnubee to be "fully supported" in mainline. Though there is still a bit of work to do, it is mostly cleaning up the code and getting it moved out of drivers/staging (spi is almost there, and mmc probably isn't far away).

This definitely deserves a **_congratulations!_** and _**Thank you!**_

**Comment by neilbrown** [comment]
Look in https://github.com/neilbrown/linux.git branch gnubee/v5.1  
This is v5.1-rc2 plus some staging patches plus a few little things from me.  A couple of changes to the DTS files are needed and there are defconfig files in there.  There are also (almost) identical config files in github.com/neilbrown/gnubee-tools.git

**Comment by neilbrown** [comment]
> The MMC driver situation is the same as the Ethernet. The mainline mtd-sd can work with it with some modifications.

Now might be the time to put that hypothesis to the test - GregKH has just removed the mt7621-mmc driver due to licensing concerns.  If you have any specific information (patches??), could you share it please.

https://lkml.org/lkml/2019/4/2/311

**Comment by neheb** [comment]
Sure. Here are some.

https://github.com/jonpry/openwrt_mt7688/commit/a85e6d99899f3dc1204cd5bfba944e17bfa6178f
https://github.com/jonpry/openwrt_mt7688/commit/24878467a650d765b747618de1a575e79114b764

A few notes: The MMC driver there is basically the 4.9 mtk-sd one with all the patches from maybe 4.17 or 4.18 backported.

**Comment by neheb** [comment]
Here's the diff that I was able to generate:

https://gist.github.com/neheb/3d9e4cbf966f8487114df19b49f28214

**Comment by neilbrown** [comment]
> Here's the diff that I was able to generate:
> 
> https://gist.github.com/neheb/3d9e4cbf966f8487114df19b49f28214

very helpful thanks.  I just booted my PC2 of the SD card using the mainline driver - with these changes and some others.  There is still polishing to do but the hard work is done.

Thanks!

**Comment by smurfix** [comment]
Is the the "boring polishing" stuff pushed to mainline by now? if not, what's left to be done?

**Comment by neilbrown** [comment]
The changes needed to drivers/mmc/host/mtd-sd.c landed in 5.2-rc1.
The changes needed to the mt7621.dtsi device tree file are in staging-next and should land in 5.3-rc1.

PCI is the main outstanding driver that needs work.  The clean-up work in staging has introduced a bug (occasional hang on boot) that no-one has found yet.

**Comment by Qwertie-** [comment]
Will mainline support mean I can download the ARM debian or fedora from their website and it will just work like it does for a desktop/laptop?

**Comment by neheb** [comment]
The GnuBee uses MIPS not ARM.

**Comment by Qwertie-** [comment]
So debian mips would work?

**Comment by vgiralt** [comment]
On 5 August 2019 15:22:19 CEST, Luke Picciau <notifications@github.com> wrote:
>So debian mips would work?

It does work, but you have to follow the instructions for installation. And I strongly recommend firmware generated using Neil Brown's tools and kernel.



-- 
Victoriano Giralt
Enviado desde el movil / Sent from mobile

**Comment by gordon-quad** [comment]
Any updates here? Looks like majority of stuff should be already merged, according to post above about 5.2 and 5.3 RCs. What about PCI driver? Is there a config that I can use to build custom mainline kernel?

**Comment by neilbrown** [comment]
In 5.17 (just released) everything except that device-tree file has been moved out of drivers/staging.
I'm running 5.15.22 on my devices using the kernel from https://github.com/neilbrown/linux/tree/gnubee/v5.15 and config file found in https://github.com/neilbrown/gnubee-tools/tree/master/kern_config
My kernel has 13 patches on top of 5.15.22.  If/when I update to 5.17 I expect that to be a lot less, but probably not zero.

**Comment by sumpfralle** [comment]
Just for my curiosity: Debian Bookworm is being released in two weeks and it will contain Linux 6.1.
Will that version be able to show the full beauty of the GnuBee?

Thanks for your work!

**Comment by neheb** [comment]
Yes
